### PR TITLE
Star tree codec changes

### PR DESCRIPTION
**Pull request by bharath-techie**
### Description
This change contains codec changes to support read / write star tree index as part of the segments.
Composite index abstraction is present at codec level to extend the codec to any future composite types.
Based on the composite index type, the codec / format will decide how to read and write.

**The actual file format changes will be followed up.** I've marked the file format related changes with todo.

### Related Issues
Resolves 
https://github.com/opensearch-project/OpenSearch/issues/14387

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/mapper/CompositeMappedFieldType.java
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/mapper/CompositeMappedFieldType.java:53]: nit: rename better, maybe data cube

**Review by Bukhtawar — COMMENTED** [review]
Should we keep it here or move it to the custom-codec repository

File: server/src/main/java/org/opensearch/index/mapper/CompositeMappedFieldType.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/mapper/CompositeMappedFieldType.java:53]: The mapping [changes](https://github.com/opensearch-project/OpenSearch/pull/14261/files) has a hierarchy - compositeMappedFieldType -> DataCubeMappedFieldType -> StarTreeMappedFieldType .. Here I've just picked this particular class to use the associated methods.

**Comment by bharath-techie** [comment]
> Should we keep it here or move it to the custom-codec repository

Since the mapping / query has first class support in core OpenSearch , and the reader needs OpenSearch specific interfaces to read the star tree data structures from index , we thought it'll be a better idea to keep the codec in core. 

Also another advantage here is that plugin codecs can still work with this Composite99Codec as their codecServiceFactory will take this as delegate.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (sarthakaggarwal97) [inline server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:95]: We can skip the creation of this map. Builder can take in the segmentWriteState, which will have the fieldInfo and maxDoc() required for the construction of star tree.

**Review by Bukhtawar — COMMENTED** [review]
Looks like it has unintended changes?

**Comment by bharath-techie** [comment]
> Looks like it has unintended changes?

Fixed it now. Looks like other commits got merged into this branch

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesFormat.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesFormat.java:62]: Won't mapper service be null here during reads (since init will happen using SPI)?

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesReader.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesReader.java:41]: Is this needed? Don't see any usage for this

Comment (mgodwan) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesReader.java:41]: Also, given MapperService can be null, not sure if this makes sense.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesReader.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesReader.java:85]: `throw new UnsupportedOperationException()` instead till we add an implementation?

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesReader.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesReader.java:92]: `throw new UnsupportedOperationException()` instead till we add an implementation?

File: server/src/main/java/org/opensearch/index/codec/CodecService.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/index/codec/CodecService.java:84]: To ensure we don't have to touch this class when any future version upgrades happen, can we get it through a factory instead which returns the latest and changes can isolated to Composite codec specific classes?

File: server/src/main/java/org/opensearch/index/codec/composite/CompositeIndexValues.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/index/codec/composite/CompositeIndexValues.java:19]: Can this be an interface?

File: server/src/main/java/org/opensearch/index/mapper/MapperService.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/index/mapper/MapperService.java:670]: How often will this be called? Is it possible to calculate this only during mapping merge so that we can return this without any active computation per call?

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:114]: [nit] `throw new UnsupporteOperationException()` until implemented?

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:77]: For StarTree it may make sense to keep this in sorted set/numeric field. For any other composite ones, do you foresee that it can be leveraged beyond these types?

**Review by mgodwan — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/index/codec/CodecService.java
Comment (sarthakaggarwal97) [inline server/src/main/java/org/opensearch/index/codec/CodecService.java:80]: I think we should put a check that only if feature flag is enabled, we will add the composite codecs.

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/CodecService.java:80]: We block in mapping phase itself based on feature flag, is it needed here ?

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:77]: For now only these 2, we'll start with sortedNumeric for now and later extend to sortedSet when we implement keyword support ?

File: server/src/main/java/org/opensearch/index/codec/CodecService.java
Comment (sarthakaggarwal97) [reply server/src/main/java/org/opensearch/index/codec/CodecService.java:80]: yeah, we should. I don't think these codecs should be loaded if the feature flag is disabled. If the feature flag is enabled, it would anyways need a restart of the process.
Also, these codecs are loaded at a shard level, its just additional objects loaded into memory.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:114]: Should we block merge ? Any composite index specific methods, we can throw exceptions , but throwing exception here also will cause actual indexing failure which will also prevent us from testing flush till merge is ready

File: server/src/main/java/org/opensearch/index/codec/CodecService.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/CodecService.java:80]: As discussed, since we have feature flag protection during index creation, we can skip adding it here.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:78]: And for other composite types, it depends on case by case basis, its a question of whether even doc values will be used, it could be extension of other formats too.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:44]: Lets call it `Composite99*` since we're introducing this format with Composite99Codec

**Review by mgodwan — APPROVED** [review]

**Review by sarthakaggarwal97 — APPROVED** [review]

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:94]: Where are we using `fieldProducerMap`? Will this be used while calling StarTree builder?

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:94]: Yes that will be needed in the star tree builder in the next set of changes. [Reference](https://github.com/opensearch-project/OpenSearch/pull/14652/files#diff-08df8016f16c9a55b9dbfd16c5360ed61d647e7b7d7f7110db7472abae5cb704R113) in merge PR

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:110]: Yes that's correct, all of these are one time operations , hence the IndexOutput once gets created will not be modified, so once merge state is set - the associated files based on IndexOutput will be based on merge flow and will get created once.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite99Codec.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite99Codec.java:31]: Yes needed for SPI. Also the reader path will be in this path.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite99Codec.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite99Codec.java:27]: Yes the delegate is PerFieldMappingPostingFormatCodec which extends Lucene99Codec for now.

So, whenever the Lucene codec gets upgraded in OpenSearch, the delegate gets changed and we need to upgrade the associated composite codec as well. We will enhance `composite codec factory` for the next version if needed.

File: server/src/main/java/org/opensearch/index/codec/composite/datacube/startree/StarTreeValues.java
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/index/codec/composite/datacube/startree/StarTreeValues.java:28]: make deep copy of the list to prevent unwanted mutation

**Review by shwetathareja — COMMENTED** [review]
Thanks @bharath-techie . Some minor comments.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:44]: Yeah everywhere I'll rename it to 99.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:50]: Yes all are handled in mapper service, it'll be empty if no composite fields are present. And in codec service, we only initialize the codec when atleast one composite field is present.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite90DocValuesWriter.java:95]: Flush is concurrent across DWPTs but IndexingChain is per DWPT.
And in each flush of DWPT, the consumer gets [initialized](https://github.com/apache/lucene/blob/62e08f5f4b76b84ed4109d8493d1dac2048d3350/lucene/core/src/java/org/apache/lucene/index/IndexingChain.java#L430) only once. Single thread handles these operations.

In merge as part of [mergeDocValues](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/index/SegmentMerger.java#L204) call , new consumer gets initialized based on new write state. So we are safe there as well.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite99Codec.java
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite99Codec.java:31]: Can we please add a comment with the use-case. IDE shows me it as un-used.

File: server/src/main/java/org/opensearch/index/codec/composite/Composite99Codec.java
Comment (reta) [inline server/src/main/java/org/opensearch/index/codec/composite/Composite99Codec.java:36]: @bharath-techie @msfroh I think we have an issue here with `PerFieldMappingPostingFormatCodec`, as per invariants, `PerFieldMappingPostingFormatCodec` always uses the latest available codec

```

    static {
        assert Codec.forName(Lucene.LATEST_CODEC).getClass().isAssignableFrom(PerFieldMappingPostingFormatCodec.class)
            : "PerFieldMappingPostingFormatCodec must subclass the latest " + "lucene codec: " + Lucene.LATEST_CODEC;
    }
```

 but `Composite99Codec` is explicitly bound to Lucene 9.9 (the latest for the current Apache Lucene version). I think we need to make a changes here while it is not too late (Apache Lucene 9.12 comes with new codecs)

Comment (mgodwan) [reply server/src/main/java/org/opensearch/index/codec/composite/Composite99Codec.java:36]: @reta I believe that should be fine. 

Lets consider the case when we upgrade to Lucene-912.

* The parametrized constructor for Composite99Codec breaks at compile time as PerFieldMappingPostingFormatCodec will accept only Lucene-912 Compression mode and the parameteized constructor in Composite99Codec will hence be removed (This is expected and preferred as we won't use this codec for any writes after upgrade).
* For writes which have happened using Composite99Codec, the reads will use the SPI constructor which is hardcoded to point to Lucene99Codec and does not cause issues.
* The new writes will happen via Composite9-12Codec which is obtained through CompositeCodecFactory as the implementation will switch to use Composite9-12Codec. The parametrized constructor will now only need to be present in this new codec.

For reference, this is how the custom-codecs plugin handles this as well.

**Pull request merged**
